tsdb/head_append.go

```diff
@@ -389,6 +400,48 @@ func (a *headAppender) Append(ref storage.SeriesRef, lset labels.Labels, t int64
 	return storage.SeriesRef(s.ref), nil
 }
 
+// AppendCreatedTimestamp appends a sample with 0 as its value.
+// AppendCreatedTimestamp does not support out-of-order appends.
+func (a *headAppender) AppendCreatedTimestamp(ref storage.SeriesRef, lset labels.Labels, t int64) (storage.SeriesRef, error) {
+	s := a.head.series.getByID(chunks.HeadSeriesRef(ref))
+	if s == nil {
+		// Ensure no empty labels have gotten through.
+		lset = lset.WithoutEmpty()
+		if lset.IsEmpty() {
+			return 0, errors.Wrap(ErrInvalidSample, "empty labelset")
+		}
+		if l, dup := lset.HasDuplicateLabelNames(); dup {
+			return 0, errors.Wrap(ErrInvalidSample, fmt.Sprintf(`label name "%s" is not unique`, l))
+		}
+		var created bool
+		var err error
+		s, created, err = a.head.getOrCreate(lset.Hash(), lset)
+		if err != nil {
+			return 0, err
+		}
+		if created {
+			a.series = append(a.series, record.RefSeries{
+				Ref:    s.ref,
+				Labels: lset,
+			})
+		}
+	}
+
+	if t > s.lastTimestamp {
+		if t > a.maxt {
+			a.maxt = t
+		}
+
+		a.samples = append(a.samples, record.RefSample{
+			Ref: s.ref,
+			T:   t,
+			V:   0.0,
+		})
+	}
+	a.sampleSeries = append(a.sampleSeries, s)
```

Comment:
I'm a bit concerned about appendable because it handles ingestion of out of order samples and I have a feeling that allowing out of order ingestion of created timestamps can become very error prone

Reply:
hmmm, reading again... appendable doesnt ingest anything, just checks if it could be ingested in-order or out-of-order...

we can simply not ingest if it is out of order

Reply:
Done, PTAL!

Reply:
> Do we need to append sample series when we don't append samples?

I don't think that's a problem... if AppendCreatedTimestamp doesn't append a sample, the next Append will since created timestamps are always exposed alongside a normal series

Reply:
Thanks, all makes sense, except that append of series for NOT ingested sample, it's super weird. Why it does not matter? and if it does not matter, why wasting cycles and confuse future code readers? (: